fedmsg/tests/crypto/test_x509.py

```diff
@@ -0,0 +1,237 @@
+# This file is part of fedmsg.
+# Copyright (C) 2012 - 2017 Red Hat, Inc.
+#
+# fedmsg is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 2.1 of the License, or (at your option) any later version.
+#
+# fedmsg is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with fedmsg; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+#
+# Authors:  Ralph Bean <rbean@redhat.com>
+# Authors:  Jeremy Cline <jcline@redhat.com>
+import os
+
+_m2crypto, _cryptography = False, False
+try:
+    import M2Crypto     # noqa: F401
+    import m2ext        # noqa: F401
+    _m2crypto = True
+except ImportError:
+    pass
+try:
+    import cryptography  # noqa
+    import OpenSSL  # noqa
+    _cryptography = True
+except ImportError:
+    pass
+
+try:
+    from unittest import skipIf, TestCase, expectedFailure
+except ImportError:
+    from unittest2 import skipIf, TestCase, expectedFailure
+
+from fedmsg import crypto  # noqa: E402
+
+
+SSLDIR = os.path.abspath(
+    os.path.join(os.path.dirname(__file__), '../test_certs/keys/'))
+
+
+@skipIf(not (_m2crypto or _cryptography), "Neither M2Crypto nor Cryptography available")
+class X509BaseTests(TestCase):
+
+    def setUp(self):
+        self.config = {
+            'ssldir': SSLDIR,
+            'certname': 'shell-app01.phx2.fedoraproject.org',
+            'ca_cert_cache': os.path.join(SSLDIR, 'ca.crt'),
+            'ca_cert_cache_expiry': 1497618475,  # Stop fedmsg overwriting my CA, See Issue 420
+
+            'crl_location': "http://threebean.org/fedmsg-tests/crl.pem",
```

Comment:
Most unfortunately, the way the code currently works it doesn't support file://. I set the cache expiry to be sometime in the mid 2030s so it should never actually try to fetch from that URL. I plan to re-write all the cert fetching code in the near future.